// ng-packagr/lib/ng-v5/entry-point/write-bundles.transform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const operators_1 = require("rxjs/operators");
const fromPromise_1 = require("rxjs/observable/fromPromise");
const pipe_1 = require("rxjs/util/pipe");
const flatten_1 = require("../../flatten/flatten");
const entry_point_node_1 = require("../entry-point.node");
exports.writeBundlesTransform = pipe_1.pipe(operators_1.switchMap(graph => {
    const entryPoint = graph.find(entry_point_node_1.byEntryPoint().and(entry_point_node_1.isInProgress));
    const ngEntryPoint = entryPoint.data.entryPoint;
    const opts = {
        entryFile: entryPoint.data.es2015EntryFile,
        outDir: entryPoint.data.stageDir,
        flatModuleFile: ngEntryPoint.flatModuleFile,
        esmModuleId: ngEntryPoint.moduleId,
        umdModuleId: ngEntryPoint.umdModuleId,
        umdModuleIds: ngEntryPoint.umdModuleIds,
        embedded: ngEntryPoint.embedded,
        comments: ngEntryPoint.comments,
        licensePath: ngEntryPoint.licensePath
    };
    return fromPromise_1.fromPromise(flatten_1.writeFlatBundleFiles(opts)).pipe(operators_1.map(() => graph));
}));
